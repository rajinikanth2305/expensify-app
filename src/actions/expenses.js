
import uuid from "uuid"
import database from '../firebase/firebase';


//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//component calls action generator
//action generator returns function
//component dispatches function
//function runs he ability to dispatch other actions and do whatever it wants




export const addExpense=(expense)=>(
    {
       type:"ADD_EXPENSE",
       expense
    }
)


export const startAddExpense = (expenseData = {}) => {
    return (dispatch,getState) => {
      const uid=getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createAt = 0
      } = expenseData;
      const expense = { description, note, amount, createAt };
  
     return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
    };
  };
  
//setup remove expense that we can dispatch to firebase
export const startRemoveExpense=({id}={})=>
{
  console.log(id);
  return (dispatch,getState)=>
  {
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>
    {
      dispatch(removeExpense({id}))
    }).catch((e)=>
    {

    })
  }
}

//we have to return keyword fro database.ref because for test case
//set up edit expense that we can dispatch to firebase


//expenses/${actions[0].expense.id}
//remove expense


export const removeExpense=({id}={})=>(
{
    type:"REMOVE_EXPENSE",
    id

}
)
//edit expense


export const editExpense=(id,updates)=>(
    {
        type:"EDIT_EXPENSE",
        id,
        updates
    }
)
export const startEditExpense=(id,updates)=>
{
  return (dispatch,getState)=>
  {
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(()=>
    {
      dispatch(editExpense(id,updates))
    })
  }
}

//SET_EXPENSES
export const setExpenses=(expenses)=>(
  {
    type:"SET_EXPENSES",
    expenses
  }
)

//fetch all expenses data once 
//pass that data tothe array
//dispatch setExpenses
//export const startSetExpenses

export const startSetExpenses=()=>
{
  return (dispatch,getState)=>
  {
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once("value").then((snapshot)=>
    {
      const expenses=[];
      snapshot.forEach((childsnapshot)=>
      {
        expenses.push({
          id:childsnapshot.key,
          ...childsnapshot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}