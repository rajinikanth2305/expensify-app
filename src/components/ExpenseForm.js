import React from "react"


import moment from "moment"
import { SingleDatePicker } from "react-dates";
//set up note state 
//const date=new Date();
//moment is standard for dates
const now = moment();
console.log(now.format("MMM Do,YYYY"))
//on change and value for text area
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            createAt: props.expense ? moment(props.expense.createAt) : now,
            calanderFocused: false,
            error: ""


        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => (
            {
                description
            }
        )
        )
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => (
            {
                note
            })
        )
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(
                {
                    amount
                }
            )
        }
    };
    onDateChange = (createAt) => {
        if (createAt) {
            this.setState({
                createAt
            })

        }
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            //set error message please provide description and amount
            this.setState(
                {
                    error: "please add description and amount"
                }
            )

        }
        else {
            this.setState(() => (
                {
                    error: ""
                }
            )
            )
            this.props.onSubmit(
                {
                    description: this.state.description,
                    amount: parseFloat(this.state.amount, 10) * 100,
                    createAt: this.state.createAt.valueOf(),
                    note: this.state.note
                }
            )
        }
    }

    render() {
        return (
            <form className="form"onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}

                <input
                    className="text-input"
                    type="text"
                    placeholder="description"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    autoFocus />

                <input
                    type="text"
                    className="text-input"

                    placeholder="amount"
                    value={this.state.amount}

                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calanderFocused}
                    onFocusChange={({ focused }) => this.setState({ calanderFocused: focused })}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className="textarea"
                    palceholder="add a note for the expense(optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}>

                </textarea>
                <div>
                <button className="button">Add expense</button>
                </div>
            </form>

        )

    }
}
//refer https://www.npmjs.com/package/react-dates for singledatepicker
//moment is used for to get an date instance and this is the best and better one
//refer react-dates
//we imported react-dates and moment library 
//moment() fucntion gives the current tiem t
//we are passing timestamp to moment() to get the date where the expense was created
//we can simulate the events and we can pass the function to that simulate and also we can set object for that event value
