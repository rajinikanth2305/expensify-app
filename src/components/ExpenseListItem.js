import React from "react";
import {NavLink} from "react-router-dom";


//export stateless functional component 
//render few pieces of information 
//description 
//amount
//creat//
const ExpenseListItem=({id,description,amount,createAt})=>(
    <div>
    <NavLink to ={`/edit/${id}`} exact={true}>
    <h1>{description}</h1>
    </NavLink>
    <p>{amount}-{createAt}</p>
    <p>{id}</p>
    </div>
 
)
export default ExpenseListItem;


//notes for this lesson

//we should have to pass expense id to edit expens epage to get edit 
//we can pass using {`/edit/${id}`} to pass id to that expense page
