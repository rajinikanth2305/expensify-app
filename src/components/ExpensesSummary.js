import React from "react"
import selectTotalExpenses from "../selectors/expense-total"
import selectExpenses from "../selectors/expenses"
import { connect } from "react-redux"
import numeral from "numeral"

export const ViewTotalExpenses=({total,length})=>
{
    const expenseWord= length===1? "expense": "expenses";
    const formattedExpenseTotal=numeral(total/100).format(('$0,0.00'));
    return(
        <div>
        <h1>Viewing {length} {expenseWord} totalling is {formattedExpenseTotal}</h1>
        
        </div>
    )
}

const mapStateToProps=(state) => {
    const visibleExpenses=selectExpenses(state.expenses,state.filters)
    return {
        total:selectTotalExpenses(visibleExpenses),
        length:visibleExpenses.length
    };
  }
export default connect(mapStateToProps)(ViewTotalExpenses)
//export default ViewTotalExpenses;