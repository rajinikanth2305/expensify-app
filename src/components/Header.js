import React from "react";
import {NavLink} from "react-router-dom";

const Header=()=>
(
    <header>
    <h1>Tilte of tha expensify app</h1>
    <NavLink to ="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
    <NavLink to ="/create" activeClassName="is-active">create expense page</NavLink>
    <NavLink to="/edit" activeClassName="is-active">Edit page</NavLink>
    <NavLink to="/help" activeClassName="is-active">Help</NavLink>

    </header>
)
export default Header;