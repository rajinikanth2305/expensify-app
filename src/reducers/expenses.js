//expenses reducer
import database from "../firebase/firebase"
const expensesReducerDefaultSate=[];

database.ref("expenses").once("value").then((snapshot)=>
{

    snapshot.forEach((childsnapshot)=>
    {
        expensesReducerDefaultSate.push(
            {
               id:childsnapshot.key,
               ...childsnapshot.val()
            }
        )
    })

})


export default (state=expensesReducerDefaultSate,action)=>
{
    switch(action.type)
    {
        case "ADD_EXPENSE":
            return [...state,action.expense]
        
        case "REMOVE_EXPENSE":
            return state.filter(({id})=>id!=action.id);
        case "EDIT_EXPENSE":
            return state.map((expense)=>
            {
                if(expense.id===action.id)
                {
                    return {...expense,...action.updates}
                }
                else{
                    return expense
                }

            })
        case "SET_EXPENSES":
            return action.expenses

        
            default:
            return state;
    }
}