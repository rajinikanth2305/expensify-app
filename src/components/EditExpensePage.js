import React from "react";
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {editExpense} from "../actions/expenses"
import {removeExpense,startEditExpense,startRemoveExpense} from "../actions/expenses"

//refractor edit expense page to be class compoenent based
//setup mapdispatch to props,editexpense and removeexpense
//should render editexpensepage
//snapshot
//shouuld handle editexpense
//spy

//should handle removeexpense
//spy
export class EditExpensePage extends React.Component{
  onSubmit=(expense)=>
  {
    this.props.startEditExpense(this.props.expense.id,expense);
    this.props.history.push("/");
  }
  onClick=()=>
  {
    const id=this.props.expense.id;
    this.props.startRemoveExpense({id});
    this.props.history.push("/")
  }
  render()
  {
    return(
      <div>
      <div className="page-header">
      <div className="content-container">
      <h1 className="page-header__title">Edit Expense</h1>
      
      </div>
      </div>
      <div className="content-container">
      <ExpenseForm onSubmit={this.onSubmit}
      expense={this.props.expense}
      />
      <button className="button button--secondary" onClick={this.onClick}>Remove</button>

      </div>
   
      </div>

    )
  }

}
const mapStateToProps = (state, props) => {
    return {
      expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
  };
const mapDispatchToProps=(dispatch,props)=>
(
  {
    startEditExpense:(id,expense)=>dispatch(startEditExpense(id,expense)),
    startRemoveExpense:(id)=>dispatch(startRemoveExpense(id))
  }
)
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);

//we can pass the current component props to mapStateToprops passing props as argument
//props.history.push is for to go to homepage
//