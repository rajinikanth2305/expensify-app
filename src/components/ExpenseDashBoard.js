import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListItem from "./ExpenseListItem"
import ExpenseListFilters from "./ExpenseListFilter"

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
