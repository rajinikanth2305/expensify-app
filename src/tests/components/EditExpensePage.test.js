import React from "react"
import {shallow} from 'enzyme'
import  expenses from "../fixtures/expenses";
import {EditExpensePage} from "../../components/EditExpensePage";
let editExpense,removeExpense,startRemoveExpense,history,wrapper;
beforeEach(()=>
{
    editExpense=jest.fn();
    startRemoveExpense=jest.fn();
    history={push:jest.fn()};
    wrapper=shallow(<EditExpensePage expense={expenses[0]} editExpense={editExpense} startRemoveExpense={startRemoveExpense} history={history} />)
}
)
test("should render edit expense page properly",()=>
{
    expect(wrapper).toMatchSnapshot();
})

//shouuld handle editexpense
test("should handle edit expense page properly",()=>
{

    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]); //passing parameter to spy function here
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,expenses[0])
})
//spy

//should handle removeexpense

test("should handle remove expense page properly",()=>
{

    wrapper.find('button').prop('onClick')(expenses[0].id);
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id:expenses[0].id})
})
//spy