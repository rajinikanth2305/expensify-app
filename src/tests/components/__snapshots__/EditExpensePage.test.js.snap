// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page properly 1`] = `
<div>
  <h1>
    paneer bill
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 800,
        "createAt": 0,
        "description": "paneer bill",
        "id": "1",
        "note": "paid by venkatesh",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
